rust/storage/src/lib.rs

```diff
@@ -94,6 +94,36 @@ impl Storage {
         }
     }
 
+    pub async fn get_parallel(&self, key: &str) -> Result<Arc<Vec<u8>>, GetError> {
+        match self {
+            Storage::S3(s3) => {
+                let res = s3.get_parallel(key).await;
+                match res {
+                    Ok(res) => Ok(res),
+                    Err(e) => match e {
+                        S3GetError::NoSuchKey(_) => Err(GetError::NoSuchKey(key.to_string())),
+                        _ => Err(GetError::S3Error(e)),
+                    },
+                }
+            }
+            Storage::Local(_) => unimplemented!(),
+            Storage::AdmissionControlledS3(admission_controlled_storage) => {
+                let res = admission_controlled_storage
+                    .get_parallel(key.to_string())
```

Comment:
why this string copy?

Reply:
need to store in the hash map for dedup'ing. Hence this can't be `&str` but `String`